File: src/struct/Set.java
package struct;

import annotation.Volatile;

import annotation.Intention;
import annotation.Required;
import lang.Class;
import nio.ForeignMemory;
import oop.TypeRegister;
import util.Hash;

import java.lang.foreign.Arena;

/**
 * Off-heap unique element hash set implementation.
 */
@Intention("Zero-GC off-heap open-addressing hash set with element class inspection, 64-bit hashing, thread-safe memory mutations, and global MemoryRegistry integration.")
@Volatile
public final class Set {

    @Required
    public static final int CLASS_ID = TypeRegister.ID_SET;

    public static final int TYPE_SET = TypeRegister.FORM_POINTER | CLASS_ID; // 0xCC000014

    private static final int DEFAULT_CAPACITY = 16;
    private static final float LOAD_FACTOR = 0.75f;
    private static final long HEADER_SIZE = 24L; // 8B metadata header + 16B slot layout
    private static final long SLOT_SIZE = 24L;   // 8B element + 8B hash + 8B state (0=empty, 1=occupied, 2=deleted)

    private static final int STATE_EMPTY = 0;
    private static final int STATE_OCCUPIED = 1;
    private static final int STATE_DELETED = 2;

    private static Arena poolArena;
    private static volatile boolean active;

    static {
        poolArena = Arena.ofShared();
        active = true;
    }

    private Set() {}

    private static void checkActive() {
        if (!active) throw new IllegalStateException("Set subsystem is not active!");
    }

    private static void checkValid(long setPtr) {
        if (setPtr == 0L) throw new NullPointerException("Accessing NULL off-heap set pointer!");
    }

    // free all subsystem resources
    public static void freeAll() {
        if (active) {
            active = false;
            if (poolArena != null && poolArena.scope().isAlive()) {
                poolArena.close();
            }
        }
    }

    // create empty off-heap set for elementClassId
    public static long instant(int elementClassId) {
        return instant(elementClassId, DEFAULT_CAPACITY);
    }

    // create empty off-heap set with initial capacity
    public static long instant(int elementClassId, int initialCapacity) {
        checkActive();
        int cap = (initialCapacity <= 0) ? DEFAULT_CAPACITY : highestOneBit(initialCapacity);
        if (cap < 4) cap = 4;

        long headerBlock = ForeignMemory.allocateNative(HEADER_SIZE);
        long userPtr = headerBlock + 8L;

        ForeignMemory.setInt(headerBlock, TYPE_SET);
        ForeignMemory.setInt(headerBlock + 4L, 0); // activeCount

        ForeignMemory.setInt(userPtr, elementClassId);
        ForeignMemory.setInt(userPtr + 4L, 0); // padding
        ForeignMemory.setInt(userPtr + 8L, cap);
        ForeignMemory.setInt(userPtr + 12L, 0); // padding

        long bufferBytes = (long) cap * SLOT_SIZE;
        long dataBuffer = ForeignMemory.allocateNative(bufferBytes);
        ForeignMemory.setMemory(dataBuffer, bufferBytes, (byte) 0);
        ForeignMemory.setLong(userPtr + 16L, dataBuffer);

        return userPtr;
    }

    // returns true if this class ID stores off-heap pointers (strings, structs) rather than raw scalar values
    private static boolean isReferenceClass(int classId) {
        return classId == TypeRegister.ID_STRING || classId >= TypeRegister.ID_LIST;
    }

    // compute 64-bit hash for element based on class inspection
    private static long computeHash(int elementClassId, long element) {
        if (element == 0L) return 0L;
        if (isReferenceClass(elementClassId) && element >= 4096L) {
            int inspectedClass = Class.getClass(element);
            if (inspectedClass != 0) {
                int len = Class.getLength(element);
                if (len > 0) return Hash.fnv1a64(element, len);
            }
        }
        return Hash.murmur3Mix64(element);
    }

    // compare two elements for equality based on class inspection
    private static boolean elementsEqual(int elementClassId, long e1, long e2) {
        if (e1 == e2) return true;
        if (e1 == 0L || e2 == 0L) return false;
        if (isReferenceClass(elementClassId) && e1 >= 4096L && e2 >= 4096L) {
            int c1 = Class.getClass(e1);
            int c2 = Class.getClass(e2);
            if (c1 != 0 && c1 == c2) {
                int len1 = Class.getLength(e1);
                int len2 = Class.getLength(e2);
                if (len1 != len2) return false;
                for (int i = 0; i < len1; i++) {
                    if (ForeignMemory.getByte(e1 + i) != ForeignMemory.getByte(e2 + i)) {
                        return false;
                    }
                }
                return true;
            }
        }
        return false;
    }


    // add element to set, returns true if inserted, false if already present
    public static synchronized boolean add(long setPtr, long element) {
        checkActive();
        checkValid(setPtr);

        int count = size(setPtr);
        int cap = capacity(setPtr);
        if (count >= (int) (cap * LOAD_FACTOR)) {
            rehash(setPtr, cap * 2);
            cap = capacity(setPtr);
        }

        int elemClassId = elementClassId(setPtr);
        long hash = computeHash(elemClassId, element);
        long buffer = dataBuffer(setPtr);
        int mask = cap - 1;
        int idx = (int) (hash & mask);
        int firstDeleted = -1;

        while (true) {
            long slot = buffer + ((long) idx * SLOT_SIZE);
            long st = ForeignMemory.getLong(slot + 16L);

            if (st == STATE_EMPTY) {
                int targetIdx = (firstDeleted != -1) ? firstDeleted : idx;
                long targetSlot = buffer + ((long) targetIdx * SLOT_SIZE);
                ForeignMemory.setLong(targetSlot, element);
                ForeignMemory.setLong(targetSlot + 8L, hash);
                ForeignMemory.setLong(targetSlot + 16L, STATE_OCCUPIED);
                ForeignMemory.setInt(setPtr - 4L, count + 1);
                return true;
            } else if (st == STATE_DELETED) {
                if (firstDeleted == -1) firstDeleted = idx;
            } else if (st == STATE_OCCUPIED) {
                long slotHash = ForeignMemory.getLong(slot + 8L);
                long slotElem = ForeignMemory.getLong(slot);
                if (slotHash == hash && elementsEqual(elemClassId, slotElem, element)) {
                    return false; // already present
                }
            }
            idx = (idx + 1) & mask;
        }
    }

    // check if set contains element
    public static synchronized boolean contains(long setPtr, long element) {
        checkValid(setPtr);
        int cap = capacity(setPtr);
        if (cap == 0) return false;

        int elemClassId = elementClassId(setPtr);
        long hash = computeHash(elemClassId, element);
        long buffer = dataBuffer(setPtr);
        int mask = cap - 1;
        int idx = (int) (hash & mask);

        for (int i = 0; i < cap; i++) {
            long slot = buffer + ((long) idx * SLOT_SIZE);
            long st = ForeignMemory.getLong(slot + 16L);

            if (st == STATE_EMPTY) return false;
            if (st == STATE_OCCUPIED) {
                long slotHash = ForeignMemory.getLong(slot + 8L);
                long slotElem = ForeignMemory.getLong(slot);
                if (slotHash == hash && elementsEqual(elemClassId, slotElem, element)) {
                    return true;
                }
            }
            idx = (idx + 1) & mask;
        }
        return false;
    }

    // remove element from set, returns true if removed, false if not found
    public static synchronized boolean remove(long setPtr, long element) {
        checkValid(setPtr);
        int cap = capacity(setPtr);
        if (cap == 0) return false;

        int elemClassId = elementClassId(setPtr);
        long hash = computeHash(elemClassId, element);
        long buffer = dataBuffer(setPtr);
        int mask = cap - 1;
        int idx = (int) (hash & mask);

        for (int i = 0; i < cap; i++) {
            long slot = buffer + ((long) idx * SLOT_SIZE);
            long st = ForeignMemory.getLong(slot + 16L);

            if (st == STATE_EMPTY) return false;
            if (st == STATE_OCCUPIED) {
                long slotHash = ForeignMemory.getLong(slot + 8L);
                long slotElem = ForeignMemory.getLong(slot);
                if (slotHash == hash && elementsEqual(elemClassId, slotElem, element)) {
                    ForeignMemory.setLong(slot + 16L, STATE_DELETED);
                    int count = size(setPtr);
                    ForeignMemory.setInt(setPtr - 4L, count - 1);
                    return true;
                }
            }
            idx = (idx + 1) & mask;
        }
        return false;
    }

    private static void rehash(long setPtr, int newCap) {
        int oldCap = capacity(setPtr);
        long oldBuffer = dataBuffer(setPtr);
        long newBufferBytes = (long) newCap * SLOT_SIZE;
        long newBuffer = ForeignMemory.allocateNative(newBufferBytes);
        ForeignMemory.setMemory(newBuffer, newBufferBytes, (byte) 0);

        int mask = newCap - 1;
        for (int i = 0; i < oldCap; i++) {
            long slot = oldBuffer + ((long) i * SLOT_SIZE);
            if (ForeignMemory.getLong(slot + 16L) == STATE_OCCUPIED) {
                long elem = ForeignMemory.getLong(slot);
                long hash = ForeignMemory.getLong(slot + 8L);

                int idx = (int) (hash & mask);
                while (ForeignMemory.getLong(newBuffer + ((long) idx * SLOT_SIZE) + 16L) == STATE_OCCUPIED) {
                    idx = (idx + 1) & mask;
                }
                long targetSlot = newBuffer + ((long) idx * SLOT_SIZE);
                ForeignMemory.setLong(targetSlot, elem);
                ForeignMemory.setLong(targetSlot + 8L, hash);
                ForeignMemory.setLong(targetSlot + 16L, STATE_OCCUPIED);
            }
        }

        ForeignMemory.freeNative(oldBuffer);
        ForeignMemory.setLong(setPtr + 16L, newBuffer);
        ForeignMemory.setInt(setPtr + 8L, newCap);
    }

    private static int highestOneBit(int i) {
        return Integer.highestOneBit(i - 1) << 1;
    }

    // export all set elements into an off-heap struct.List
    public static synchronized long toList(long setPtr) {
        checkValid(setPtr);
        int elemClassId = elementClassId(setPtr);
        int count = size(setPtr);
        long listPtr = List.instant(elemClassId, count);
        if (count == 0) return listPtr;

        int cap = capacity(setPtr);
        long buffer = dataBuffer(setPtr);
        for (int i = 0; i < cap; i++) {
            long slot = buffer + ((long) i * SLOT_SIZE);
            if (ForeignMemory.getLong(slot + 16L) == STATE_OCCUPIED) {
                long elem = ForeignMemory.getLong(slot);
                List.add(listPtr, elem);
            }
        }
        return listPtr;
    }

    // export all set elements into a sorted off-heap struct.List
    public static synchronized long toSortedList(long setPtr) {
        long listPtr = toList(setPtr);
        int sz = List.size(listPtr);
        if (sz > 1) {
            long buf = List.dataBuffer(listPtr);
            int stride = List.stride(listPtr);
            if (stride == 4) {
                util.Arrays.sortInt(buf, sz);
            } else if (stride == 8) {
                util.Arrays.sortLong(buf, sz);
            }
        }
        return listPtr;
    }


    // check if set is empty
    public static boolean isEmpty(long setPtr) {
        return Collection.isEmpty(setPtr);
    }

    // get active element count
    public static int size(long setPtr) {
        return Collection.size(setPtr);
    }

    // free set data buffer and header back to native RAM
    public static synchronized void free(long setPtr) {
        checkActive();
        if (setPtr == 0L) return;

        long headerBlock = setPtr - 8L;
        int type = ForeignMemory.getInt(headerBlock);
        if (type == 0 || !TypeRegister.isPointer(type)) {
            throw new IllegalStateException("Double free or corrupt set pointer: 0x" + Long.toHexString(setPtr).toUpperCase());
        }

        long buffer = dataBuffer(setPtr);
        if (buffer != 0L) {
            ForeignMemory.freeNative(buffer);
        }

        ForeignMemory.setInt(headerBlock, 0);
        ForeignMemory.setInt(headerBlock + 4L, -1);
        ForeignMemory.freeNative(headerBlock);
    }

    public static int elementClassId(long setPtr) {
        return Collection.elementClassId(setPtr);
    }

    public static int capacity(long setPtr) {
        return Collection.capacity(setPtr);
    }

    public static long dataBuffer(long setPtr) {
        return Collection.dataBuffer(setPtr);
    }

    public static int classId() {
        return CLASS_ID;
    }

}

File: src/struct/set.c
#include "struct/set.h"

#include <string.h>

#include "nio/mem.h"
#include "oop/type.h"
#include "util/arrays.h"
#include "util/hash.h"
#include "annotation/overview.h"

;;OVERVIEW
/**
 * ============================================================================
 * CLASS: Set (struct/set.c)
 * LEVEL: L2 — Behavior (container behavior API)
 * ============================================================================
 * the Set class, ported from struct/Set.java.
 *
 * STRUCT FIELDS (Mirroring struct/set.h):
 * ----------------------------------------------------------------------------
 *   Set {
 *     Collection collection; // instance state
 *   }
 *
 * FUNCTION REGISTRY:
 * ----------------------------------------------------------------------------
 * Constructors:
 *   - Set_2(element_class, capacity)
 *   - Set_1(element_class)
 *
 * Setters:
 *   - Set_free(set)
 *   - Set_add(set, element)
 *   - Set_contains(set, element)
 *   - Set_remove(set, element)
 *   - Set_toList(set)
 *   - Set_toSortedList(set)
 *   - Set_isEmpty(set)
 *   - Set_size(set)
 *   - Set_capacity(set)
 *   - Set_elementClassId(set)
 *   - Set_dataBuffer(set)
 * ============================================================================
 */


// set.c — Set port (Legacy: struct/Set.java).

static const size_t SLOT_SIZE = 24;
static const size_t DEFAULT_CAPACITY = 16;
static const size_t LOAD_DIVISOR = 4; // 0.75 load factor
static const uint64_t STATE_EMPTY = 0;
static const uint64_t STATE_OCCUPIED = 1;
static const uint64_t STATE_DELETED = 2;

static Collection *asCollection(Set *set) {
    return (Collection*) set;
}

static int isReferenceClass(uint32_t classId) {
    return classId == ID_STRING || classId >= ID_LIST;
}

static uint64_t computeHash(uint32_t elementClass, uint64_t element) {
    if (element == 0) return 0;
    if (isReferenceClass(elementClass) && element >= 4096u) {
        uint64_t inspected = Memory_type((void*) (uintptr_t)element);
        if (inspected != 0) {
            size_t len = Memory_length((void*) (uintptr_t)element);
            if (len > 0)
                return Hash_fnv1a64((const uint8_t*) (uintptr_t)element, len);
        }
    }
    return Hash_murmur3Mix64(element);
}

static int elementsEqual(uint32_t elementClass, uint64_t e1, uint64_t e2) {
    if (e1 == e2) return 1;
    if (e1 == 0 || e2 == 0) return 0;
    if (isReferenceClass(elementClass) && e1 >= 4096u && e2 >= 4096u) {
        void *p1 = (void*) (uintptr_t)e1;
        void *p2 = (void*) (uintptr_t)e2;
        uint64_t t1 = Memory_type(p1);
        uint64_t t2 = Memory_type(p2);
        if (t1 != 0 && t1 == t2) {
            size_t len1 = Memory_length(p1);
            size_t len2 = Memory_length(p2);
            if (len1 != len2) return 0;
            return memcmp(p1, p2, len1) == 0;
        }
    }
    return 0;
}

static size_t highestOneBit(size_t n) {
    if (n <= 1) return 1;
    size_t v = n - 1;
    v |= v >> 1;
    v |= v >> 2;
    v |= v >> 4;
    v |= v >> 8;
    v |= v >> 16;
    v |= v >> 32;
    return v + 1;
}

static uint8_t *slotAt(Collection *c, size_t index) {
    return (*c).data + index * SLOT_SIZE;
}

static uint64_t slotState(uint8_t *slot) {
    return *(uint64_t*) (slot + 16);
}

static void rehash(Collection *c, size_t newCap) {
    size_t oldCap = (*c).capacity;
    uint8_t *oldData = (*c).data;
    size_t bytes = newCap * SLOT_SIZE;
    uint64_t bufType = Type_make(PROJ_VEXSPOKE, FORM_ARRAY, ID_SET);
    uint8_t *newData = (uint8_t*) Memory_alloc(bufType, bytes);
    if (!newData) return;
    memset(newData, 0, bytes);

    size_t mask = newCap - 1;
    for (size_t i = 0; i < oldCap; i++) {
        uint8_t *slot = slotAt(c, i);
        if (slotState(slot) == STATE_OCCUPIED) {
            uint64_t elem = *(uint64_t*) slot;
            uint64_t hash = *(uint64_t*) (slot + 8);

            size_t idx = (size_t)(hash & mask);
            while (slotState(newData + idx * SLOT_SIZE) == STATE_OCCUPIED)
                idx = (idx + 1) & mask;
            uint8_t *target = newData + idx * SLOT_SIZE;
            *(uint64_t*) target = elem;
            *(uint64_t*) (target + 8) = hash;
            *(uint64_t*) (target + 16) = STATE_OCCUPIED;
        }
    }

    Memory_free(oldData);
    (*c).data = newData;
    (*c).capacity = (uint32_t)newCap;
}

Set *Set_2(uint32_t elementClass, size_t capacity) {
    size_t cap = capacity == 0 ? DEFAULT_CAPACITY : highestOneBit(capacity);
    if (cap < 4) cap = 4;

    Set *set = (Set*) Memory_alloc(TYPE_SET, sizeof(Set));
    if (!set) return nullptr;

    Collection *c = asCollection(set);
    (*c).typeId = TYPE_SET;
    (*c).activeCount = 0;
    (*c).elementClass = elementClass;
    (*c).stride = 0;
    (*c).capacity = (uint32_t)cap;
    (*c).head = 0;

    size_t bytes = cap * SLOT_SIZE;
    uint64_t bufType = Type_make(PROJ_VEXSPOKE, FORM_ARRAY, ID_SET);
    (*c).data = (uint8_t*) Memory_alloc(bufType, bytes);
    if (!(*c).data) {
        Memory_free(set);
        return nullptr;
    }
    memset((*c).data, 0, bytes);
    return set;
}

void Set_free(Set *set) {
    if (!set) return;
    Collection *c = asCollection(set);
    if ((*c).data)
        Memory_free((*c).data);
    Memory_free(set);
}

int Set_add(Set *set, uint64_t element) {
    if (!set) return 0;
    Collection *c = asCollection(set);

    size_t load = (*c).capacity - (*c).capacity / LOAD_DIVISOR;
    if ((*c).activeCount >= load)
        rehash(c, (*c).capacity * 2);

    uint64_t hash = computeHash((*c).elementClass, element);
    size_t mask = (*c).capacity - 1;
    size_t idx = (size_t)(hash & mask);
    size_t firstDeleted = SIZE_MAX;

    while (true) {
        uint8_t *slot = slotAt(c, idx);
        uint64_t st = slotState(slot);

        if (st == STATE_EMPTY) {
            size_t target = firstDeleted != SIZE_MAX ? firstDeleted : idx;
            uint8_t *tslot = slotAt(c, target);
            *(uint64_t*) tslot = element;
            *(uint64_t*) (tslot + 8) = hash;
            *(uint64_t*) (tslot + 16) = STATE_OCCUPIED;
            (*c).activeCount++;
            return 1;
        } else if (st == STATE_DELETED) {
            if (firstDeleted == SIZE_MAX)
                firstDeleted = idx;
        } else if (st == STATE_OCCUPIED) {
            if (*(uint64_t*) (slot + 8) == hash
                && elementsEqual((*c).elementClass, *(uint64_t*) slot, element))
                return 0;
        }
        idx = (idx + 1) & mask;
    }
}

bool Set_contains(Set *set, uint64_t element) {
    if (!set) return false;
    Collection *c = asCollection(set);
    if ((*c).capacity == 0) return false;

    uint64_t hash = computeHash((*c).elementClass, element);
    size_t mask = (*c).capacity - 1;
    size_t idx = (size_t)(hash & mask);

    for (size_t i = 0; i < (*c).capacity; i++) {
        uint8_t *slot = slotAt(c, idx);
        uint64_t st = slotState(slot);
        if (st == STATE_EMPTY) return false;
        if (st == STATE_OCCUPIED
            && *(uint64_t*) (slot + 8) == hash
            && elementsEqual((*c).elementClass, *(uint64_t*) slot, element))
            return true;
        idx = (idx + 1) & mask;
    }
    return false;
}

int Set_remove(Set *set, uint64_t element) {
    if (!set) return 0;
    Collection *c = asCollection(set);
    if ((*c).capacity == 0) return 0;

    uint64_t hash = computeHash((*c).elementClass, element);
    size_t mask = (*c).capacity - 1;
    size_t idx = (size_t)(hash & mask);

    for (size_t i = 0; i < (*c).capacity; i++) {
        uint8_t *slot = slotAt(c, idx);
        uint64_t st = slotState(slot);
        if (st == STATE_EMPTY) return 0;
        if (st == STATE_OCCUPIED
            && *(uint64_t*) (slot + 8) == hash
            && elementsEqual((*c).elementClass, *(uint64_t*) slot, element)) {
            *(uint64_t*) (slot + 16) = STATE_DELETED;
            (*c).activeCount--;
            return 1;
        }
        idx = (idx + 1) & mask;
    }
    return 0;
}

List *Set_toList(Set *set) {
    if (!set) return nullptr;
    Collection *c = asCollection(set);
    List *list = List((*c).elementClass, (*c).activeCount);
    if (!list) return nullptr;

    for (size_t i = 0; i < (*c).capacity; i++) {
        uint8_t *slot = slotAt(c, i);
        if (slotState(slot) == STATE_OCCUPIED)
            List_add(list, *(uint64_t*) slot);
    }
    return list;
}

List *Set_toSortedList(Set *set) {
    List *list = Set_toList(set);
    if (!list) return nullptr;
    if (List_size(list) > 1) {
        size_t stride = List_stride(list);
        uint8_t *buf = List_dataBuffer(list);
        size_t n = List_size(list);
        if (stride == 4)
            Arrays_sortInt((int32_t*) buf, n);
        else if (stride == 8)
            Arrays_sortLong((int64_t*) buf, n);
    }
    return list;
}

bool Set_isEmpty(Set *set) {
    return Collection_isEmpty(asCollection(set));
}

size_t Set_size(Set *set) {
    return Collection_size(asCollection(set));
}

size_t Set_capacity(Set *set) {
    return Collection_capacity(asCollection(set));
}

uint32_t Set_elementClassId(Set *set) {
    return Collection_elementClassId(asCollection(set));
}

uint8_t *Set_dataBuffer(Set *set) {
    return Collection_dataBuffer(asCollection(set));
}
Set *Set_1(uint32_t element_class) {
    return Set_2(element_class, 16);
}
